src/relay/pass/pass_util.h

```diff
@@ -108,6 +109,54 @@ inline bool IsAtomic(const Expr& e) {
   return e.as<VarNode>() || e.as<OpNode>() || e.as<ConstructorNode>() || e.as<GlobalVarNode>();
 }
 
+struct TreeNode {
+  virtual ~TreeNode() {}
+};
+using TreeNodePtr = std::shared_ptr<TreeNode>;
+
+struct TreeLeafNode : TreeNode {
+  Expr body;
+
+  explicit TreeLeafNode(Expr body): body(body) {}
+
+  static TreeNodePtr Make(Expr body) {
+    return std::make_shared<TreeLeafNode>(body);
+  }
+
+  ~TreeLeafNode() {}
+};
+
+struct TreeLeafFatalNode : TreeNode {
+  TreeLeafFatalNode() = default;
+
+  static TreeNodePtr Make() {
+    return std::make_shared<TreeLeafFatalNode>();
+  }
+
+  ~TreeLeafFatalNode() {}
+};
+
+template<typename ConditionNodePtr>
```

Comment:
move the template to TreeNode. We do not want treenode to contain arbitary type of conditionnodeptr, but whether a single type specified beforehand.